File: fb/246.Strobogrammatic-Number.cpp
// Time O(n)
// Space O(1)
class Solution {
public:
    bool isStrobogrammatic(string num) {
        const unordered_map<char, char> um = {{'0', '0'},
                                              {'1', '1'},
                                              {'6', '9'},
                                              {'8', '8'},
                                              {'9', '6'}};
        for(int i = 0; i < (num.size() + 1) / 2; i++) {
            if(!um.count(num[i])) 
                return false;
            if(um.at(num[i]) != num[num.size()-1-i]) 
                return false;
        }
        return true;
    }
};

File: fb/246.Strobogrammatic-Number.py
class Solution:
    def isStrobogrammatic(self, num: str) -> bool:
        mapper = {"0": "0", "1": "1", "6": "9", "8": "8", "9": "6"}
        left = 0
        right = len(num)-1
        
        while left <= right:
            if num[left] not in mapper or mapper[num[left]] != num[right]:
                return False
            left += 1
            right -= 1
            
        return True
